File: dsa-problems/LinkedLists/solutions/python/LNK-006-lab-loop-detector-entry-length.py
import sys

class ListNode:
    def __init__(self, val=0):
        self.val = val
        self.next = None

def cycle_info(head: ListNode):
    if not head:
        return (-1, 0, 0)
    
    slow = head
    fast = head
    has_cycle = False
    
    # Phase 1: Detect
    while fast and fast.next:
        slow = slow.next
        fast = fast.next.next
        if slow == fast:
            has_cycle = True
            break
            
    if not has_cycle:
        return (-1, 0, 0)
        
    # Phase 2: Find Entry
    entry = head
    entry_index = 0
    while entry != slow:
        entry = entry.next
        slow = slow.next
        entry_index += 1
        
    # Phase 3: Stats
    length = 0
    max_val = -float('inf')
    curr = entry
    while True:
        length += 1
        max_val = max(max_val, curr.val)
        curr = curr.next
        if curr == entry:
            break
            
    return (entry_index, length, max_val)

def main():
    input = sys.stdin.read
    data = input().split()
    if not data:
        return
    
    iterator = iter(data)
    try:
        n = int(next(iterator))
        dummy = ListNode()
        cur = dummy
        nodes = []
        for _ in range(n):
            node = ListNode(int(next(iterator)))
            cur.next = node
            cur = cur.next
            nodes.append(node)
            
        pos = int(next(iterator))
        if pos >= 0 and n > 0:
            cur.next = nodes[pos]
            
        entry, length, max_val = cycle_info(dummy.next)
        print(f"{entry} {length} {max_val}")
    except StopIteration:
        pass

if __name__ == "__main__":
    main()

File: dsa-problems/LinkedLists/solutions/java/LNK-006-lab-loop-detector-entry-length.java
import java.util.*;

class ListNode {
    int val;
    ListNode next;
    ListNode(int val) { this.val = val; }
}

class Solution {
    public int[] cycleInfo(ListNode head) {
        if (head == null) return new int[]{-1, 0, 0};

        ListNode slow = head;
        ListNode fast = head;
        boolean hasCycle = false;

        // Phase 1: Detect Cycle
        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
            if (slow == fast) {
                hasCycle = true;
                break;
            }
        }

        if (!hasCycle) return new int[]{-1, 0, 0};

        // Phase 2: Find Entry
        ListNode entry = head;
        int entryIndex = 0;
        while (entry != slow) {
            entry = entry.next;
            slow = slow.next;
            entryIndex++;
        }

        // Phase 3: Cycle Stats
        int length = 0;
        int maxVal = Integer.MIN_VALUE;
        ListNode curr = entry;
        do {
            length++;
            maxVal = Math.max(maxVal, curr.val);
            curr = curr.next;
        } while (curr != entry);

        return new int[]{entryIndex, length, maxVal};
    }
}

class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        if (!sc.hasNextInt()) return;
        int n = sc.nextInt();
        
        ListNode dummy = new ListNode(0);
        ListNode cur = dummy;
        ListNode[] nodes = new ListNode[n];
        for (int i = 0; i < n; i++) {
            cur.next = new ListNode(sc.nextInt());
            cur = cur.next;
            nodes[i] = cur;
        }
        
        int pos = sc.nextInt();
        if (pos >= 0 && n > 0) {
            cur.next = nodes[pos];
        }

        Solution solution = new Solution();
        int[] res = solution.cycleInfo(dummy.next);
        System.out.println(res[0] + " " + res[1] + " " + res[2]);
        sc.close();
    }
}
